### ISSUE TITLE
这里操作是不是有问题？

### ISSUE THREAD

**Issue by pkuyilong** [xiongzihua/pytorch-YOLO-v1#33]
```
box1_xyxy[:, :2] = box1[:, :2] / 14. - 0.5 * box1[:, 2:4]
box1_xyxy[:, 2:4] = box1[:, :2] / 14. + 0.5 * box1[:, 2:4]
box2 = box_target[i].view(-1, 5)
box2_xyxy = Variable(torch.FloatTensor(box2.size()))
box2_xyxy[:, :2] = box2[:, :2] / 14. - 0.5 * box2[:, 2:4]
box2_xyxy[:, 2:4] = box2[:, :2] / 14. + 0.5 * box2[:, 2:4]
```
这里预测出来的xywh应该都是[0-1]，这里除以14没有意义吧

**Comment by pkuyilong** [comment]
@xiongzihua 如果有空恳请交流一下。

**Comment by abeardear** [comment]
@pkuyilong 
box1[:, :2]是在0-1之间，预测的是相对该网格左上角的x1,y1，是相对于网格的0-1。
box1[:, 2:4]也在0-1，预测的是wh，是相对全图的0-1。
计算IOU时，除以14将x1,y1的坐标从网格视角转换到全图视角，不知道这样能不能理解，代码的意图就是这样的，不知道是否合理，你可以继续思考一下，希望对你有所帮助

**Comment by HusterRC** [comment]
@pkuyilong do you understand the idea of the author?

**Comment by HusterRC** [comment]
@xiongzihua 能举个例子说明这种计算的正确性吗？

**Comment by pkuyilong** [comment]
> @pkuyilong do you understand the idea of the author?

Training phrase:
the prediction generated from model is 14 x 14,
the ground truth is scaled by origin image size(maybe 480 x 480),
so when calculating the loss, we need to divide the prediction data by 14, which is equal to scale the prediction data by origin image size.

**Comment by pkuyilong** [comment]
> @pkuyilong
> box1[:, :2]是在0-1之间，预测的是相对该网格左上角的x1,y1，是相对于网格的0-1。
> box1[:, 2:4]也在0-1，预测的是wh，是相对全图的0-1。
> 计算IOU时，除以14将x1,y1的坐标从网格视角转换到全图视角，不知道这样能不能理解，代码的意图就是这样的，不知道是否合理，你可以继续思考一下，希望对你有所帮助

又看了一会代码，差不多可以理解了，谢谢🙏

**Comment by HusterRC** [comment]
@pkuyilong  我测试了一下网络，发现预测的IOU较小，没有box。自己实现了一下IOU计算，以下是代码
index = 0
        for i_batchsize in range(N):
            for i in range(self.S):
                for j in range(self.S):
                    if target_tensor[i_batchsize, i, j, 4] == 1:   # 说明这里需要计算IOU
                        box_gt = target_tensor[i_batchsize, i, j, 0:4].cpu().numpy()
                        box_gt[0] += (j *1.0/ 14)
                        box_gt[1] += (i * 1.0/ 14)
                        box1_pre = pred_tensor[i_batchsize, i, j, 0:4].cpu().numpy()
                        box1_pre[0] += (j * 1.0/14)
                        box1_pre[0] += (i * 1.0/14)

                        box2_pre = pred_tensor[i_batchsize, i, j, 5:9].cpu().numpy()
                        box2_pre[0] += (j * 1.0/14)
                        box2_pre[0] += (i * 1.0/14)
                        box_pre = np.concatenate((box1_pre, box2_pre), axis=0)
                        box_gt_IOU = torch.from_numpy(my_decode(box_gt))
                        box_pre_IOU = torch.from_numpy(my_decode(box_pre))

                        iou = self.compute_iou( box_gt_IOU,box_pre_IOU) #[2,1]
                        max_iou,max_index = iou.max(0)
                        coo_response_mask[index+max_index]=1
                        coo_not_response_mask[index+1-max_index]=1
                        box_target_iou[index+max_index, 4] = max_iou
                        index +=2
        assert(index == box_target.shape[0])

**Comment by HusterRC** [comment]
@pkuyilong @xiongzihua 我自己实现了以下代码，你这种方式始终感觉不合适，计算的IOU值肯定与实际的IOU值有区别，以下是实现的代码。
index = 0
        for i_batchsize in range(N):
            for i in range(self.S):
                for j in range(self.S):
                    if target_tensor[i_batchsize, i, j, 4] == 1:   # 说明这里需要计算IOU
                        box_gt = target_tensor[i_batchsize, i, j, 0:4].cpu().numpy()
                        box_gt[0] += (j *1.0/ 14)
                        box_gt[1] += (i * 1.0/ 14)
                        box1_pre = pred_tensor[i_batchsize, i, j, 0:4].cpu().numpy()
                        box1_pre[0] += (j * 1.0/14)
                        box1_pre[1] += (i * 1.0/14)

                        box2_pre = pred_tensor[i_batchsize, i, j, 5:9].cpu().numpy()
                        box2_pre[0] += (j * 1.0/14)
                        box2_pre[1] += (i * 1.0/14)
                        box_pre = np.concatenate((box1_pre, box2_pre), axis=0)
                        box_gt_IOU = torch.from_numpy(my_decode(box_gt))
                        box_pre_IOU = torch.from_numpy(my_decode(box_pre))

                        iou = self.compute_iou( box_gt_IOU,box_pre_IOU) #[2,1]
                        max_iou,max_index = iou.max(0)
                        coo_response_mask[index+max_index]=1
                        coo_not_response_mask[index+1-max_index]=1
                        box_target_iou[index+max_index, 4] = max_iou
                        index +=2
        assert(index == box_target.shape[0])
